### PR TITLE
TST Adds module test comparing output of cpu and gpu

### PR DESCRIPTION
**Pull request by thomasjpfan**
Continues #61935

This PR adds:

1. `test_cpu_gpu_results_are_equal`, which checks the results from cpu and cuda.
2. `allow_devices` to `modules` to filter tests without skipping.

Note this PR does not compare gradgrad check to keep the PR smaller.

cc @albanD @mruberry @jbschlosser @walterddr

### CODE REVIEW

File: torch/testing/_internal/common_modules.py [hunk 1]
```diff
@@ -176,13 +179,14 @@ def module_inputs_torch_nn_Linear(module_info, device, dtype, requires_grad, **k
...
 
     cases: List[Tuple[str, dict]] = [
         ('', {}),
         ('ignore_index', {'ignore_index': 2}),
-        ('weights', {'weight': make_input(10)}),
-        ('weights_ignore_index', {'weight': make_input(10), 'ignore_index': 2}),
-        ('weights_ignore_index_neg', {'weight': make_input(10), 'ignore_index': -1})
+        ('weights', {'weight': make_weight(10).abs()}),
+        ('weights_ignore_index', {'weight': make_weight(10).abs(), 'ignore_index': 2}),
+        ('weights_ignore_index_neg', {'weight': make_weight(10).abs(), 'ignore_index': -1})
```
Comment (thomasjpfan) [inline torch/testing/_internal/common_modules.py:189]: This diff is in every module test PR that checks the gradients.

File: torch/testing/_internal/common_modules.py [hunk 2]
```diff
@@ -159,8 +162,8 @@ def module_inputs_torch_nn_Linear(module_info, device, dtype, requires_grad, **k
 
     module_inputs = [
         ModuleInput(constructor_input=FunctionInput(10, 8),
-                    forward_input=FunctionInput(make_input((4, 10))),
-                    reference_fn=lambda m, p, i: torch.mm(i, p[0].t()) + p[1].view(1, -1).expand(4, 8)),
+                    forward_input=FunctionInput(input=make_input((4, 10))),
+                    reference_fn=lambda m, p, input: torch.mm(input, p[0].t()) + p[1].view(1, -1).expand(4, 8)),
```
Comment (thomasjpfan) [inline torch/testing/_internal/common_modules.py:166]: Small check to make sure kwargs is supported.

File: test/test_modules.py [hunk 3]
```diff
@@ -156,6 +156,95 @@ def test_pickle(self, device, dtype, module_info):
                     output_from_copy = m_copy(*args, **kwargs)
                     self.assertEqual(output, output_from_copy)
 
+    @modules(module_db, allow_devices={"cuda"})
+    def test_cpu_gpu_results_are_equal(self, device, dtype, module_info):
+        # Test cpu and gpu results are the same
+        module_cls = module_info.module_cls
+        module_inputs_cpu = module_info.module_inputs_func(module_info, device="cpu", dtype=dtype,
+                                                           requires_grad=True)
```
Comment (thomasjpfan) [inline test/test_modules.py:164]: I think it's a little more "standard" to create the tensors on a cpu and move to a gpu, but I am okay either way.

File: test/test_modules.py [hunk 4]
```diff
@@ -156,6 +156,95 @@ def test_pickle(self, device, dtype, module_info):
...
+
+        def make_leafs_input(items):
```
Comment (thomasjpfan) [inline test/test_modules.py:166]: This is for cases where the input are not leaves. For example in NLLLoss, the forward_input args are not leaves:

```python
import torch
from torch.testing import make_tensor

input = make_tensor((15, 10), device='cpu', dtype=torch.float32, requires_grad=True).log_softmax(dim=1)
input.is_leaf
# False
```

Comment (albanD) [reply test/test_modules.py]: WHy do you need them to be leafs?
And the only way for them not to be leafs is if they do require gradients so not a problem for gradcheck

File: test/test_modules.py [hunk 5]
```diff
@@ -156,6 +156,94 @@ def test_pickle(self, device, dtype, module_info):
...
+            if isinstance(item, dict):
+                for i in item.values():
+                    _make_leafs(i)
+            elif isinstance(item, tuple):
+                for i in item:
+                    _make_leafs(i)
+            else:
+                if not isinstance(item, torch.Tensor) or item.is_leaf:
+                    return
+                old_requires_grad = item.requires_grad
```
Comment (jbschlosser) [inline test/test_modules.py:176]: Hm, this is interesting. Is this needed because of NLLLoss's ModuleInputs? Looks like the input there is created with `requires_grad=True` then `log_softmax()` is called, creating a graph which breaks the test.

I wonder if we want this somewhere more general to enforce that all tensors with `requires_grad=True` coming from `module_inputs_func`s are guaranteed to have an empty graph. Wouldn't this be needed for the gradcheck tests as well?

Also if we do make it more general and called automagically, it probably should handle lists as well.

**Review by jbschlosser — COMMENTED** [review]
Looking pretty good, thanks for the update! Few comments below

File: test/test_modules.py (hunk 4 above)
Comment (thomasjpfan) [reply test/test_modules.py]: This PR is checking for cpu+gpu parity for `_.grad`. `backward()` only accumulates gradients in the leafs:

```python
import torch
from torch.testing import make_tensor
import torch.nn as nn

input = (make_tensor((15, 10), device='cpu', dtype=torch.float32, requires_grad=True)
         .log_softmax(dim=1))
print("input.requires_grad:", input.requires_grad)

target = torch.empty(15, device='cpu').uniform_().mul(10).floor().long()

nll = nn.NLLLoss()
loss = nll(input, target)
loss.backward()
print("input.grad:", input.grad)

# input.requires_grad: True
# input.grad: None
```

This also issues a `UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed.`

File: test/test_modules.py (hunk 5 above)
Comment (thomasjpfan) [reply test/test_modules.py]: > Wouldn't this be needed for the gradcheck tests as well?

`gradcheck` will check the input and call `retain_grad`:

https://github.com/pytorch/pytorch/blob/c12df2dc2358e58c1dc66221e18635fe29f19e06/torch/autograd/gradcheck.py#L663

I updated this function to do the same.

File: test/test_modules.py [hunk 6]
```diff
@@ -156,6 +156,94 @@ def test_pickle(self, device, dtype, module_info):
                     output_from_copy = m_copy(*args, **kwargs)
                     self.assertEqual(output, output_from_copy)
 
+    @modules(module_db, allow_devices={"cuda"})
```
[the remark this answers is not in the document]
Comment (thomasjpfan) [reply test/test_modules.py]: As long as we are okay with skipped tests, `@onlyCUDA` works.  I updated PR with `@onlyCUDA`.

File: test/test_modules.py [hunk 7]
```diff
@@ -156,6 +156,94 @@ def test_pickle(self, device, dtype, module_info):
...
+            # === Construct module on cpu and gpu ===
+            args, kwargs = module_input.constructor_input.args, module_input.constructor_input.kwargs
+
+            cpu_module = module_cls(*args, **kwargs).to(dtype).to("cpu")
+            gpu_module = module_cls(*args, **kwargs).to(dtype).to(device)
+
+            for cpu_p, gpu_p in zip(cpu_module.parameters(), gpu_module.parameters()):
+                gpu_p.data.copy_(cpu_p)
+
+            # === Compare forward resputs for cpu and gpu ===
```
[the remark this answers is not in the document]
Comment (thomasjpfan) [reply test/test_modules.py]: (it's a typo) :)

File: test/test_modules.py [hunk 8]
```diff
@@ -156,6 +156,94 @@ def test_pickle(self, device, dtype, module_info):
...
+            if isinstance(obj, torch.Tensor):
+                res = obj.detach().to(device=device)
+                res.requires_grad = obj.requires_grad
+                return res
+            elif isinstance(obj, tuple):
+                return tuple(_to_device(o) for o in obj)
+            elif isinstance(obj, dict):
+                return {key: _to_device(o) for key, o in obj.items()}
+            else:
+                return deepcopy(obj)
```
[the remark this answers is not in the document]
Comment (thomasjpfan) [reply test/test_modules.py:195]: I think freezing rng does not give the same tensors between cpu and gpu:

```python
import torch
from torch.testing._internal.common_utils import freeze_rng_state
from torch.testing import make_tensor

with freeze_rng_state():
    x_cpu = make_tensor((2, 3), device="cpu", dtype=torch.float32)
with freeze_rng_state():
    x_gpu = make_tensor((2, 3), device="cuda", dtype=torch.float32)

torch.allclose(x_cpu, x_gpu.cpu())
# False
```

File: test/test_modules.py (hunk 4 above)
Comment (albanD) [reply test/test_modules.py]: In that case, 
```python
for item in items:
    if isinstance(item, torch.Tensor):
        item.retain_grad()
```

File: test/test_modules.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_modules.py:195]: Eek fair enough, disregard my suggestion :)

File: test/test_modules.py [hunk 9]
```diff
@@ -156,6 +158,95 @@ def test_pickle(self, device, dtype, module_info):
...
+            cpu_module = module_cls(*args, **kwargs).to(dtype).to("cpu")
+            gpu_module = module_cls(*args, **kwargs).to(dtype).to(device)
+
+            for cpu_p, gpu_p in zip(cpu_module.parameters(), gpu_module.parameters()):
+                gpu_p.data.copy_(cpu_p)
+
+            # === Compare forward output between cpu and gpu ===
+            with freeze_rng_state():
+                cpu_output = cpu_module(*cpu_forward_args, **cpu_forward_kwargs)
+            with freeze_rng_state():
```
Comment (jbschlosser) [inline test/test_modules.py:219]: nit: I don't think you need this second `freeze_rng_state` call

File: test/test_modules.py [hunk 10]
```diff
@@ -156,6 +158,95 @@ def test_pickle(self, device, dtype, module_info):
...
+    @toleranceOverride({torch.float32: tol(5e-2, 0),
+                        torch.float64: tol(4e-4, 0)})
+    @modules(module_db)
+    def test_cpu_gpu_parity(self, device, dtype, module_info):
+        # Test cpu and gpu results are the same
+        module_cls = module_info.module_cls
+        module_inputs_cpu = module_info.module_inputs_func(module_info, device="cpu", dtype=dtype,
+                                                           requires_grad=True)
+
+        def _retain_grad(item):
```
Comment (jbschlosser) [inline test/test_modules.py:171]: Do you mind adding a comment explaining the reason this is necessary?

**Comment by thomasjpfan** [comment]
Depending on what the issue is in https://github.com/pytorch/pytorch/pull/64444#issuecomment-917677164, this PR may also cause issues.

**Comment by jbschlosser** [comment]
> Depending on what the issue is in [#64444 (comment)](https://github.com/pytorch/pytorch/pull/64444#issuecomment-917677164), this PR may also cause issues.

@thomasjpfan FYI a `@skipIfTBB` decorator is being added in #64942 to address the failures

**Comment by thomasjpfan** [comment]
@jbschlosser I wonder if that is why the number of threads is set to one in `common_nn`:

https://github.com/pytorch/pytorch/blob/0561e104d97cb2abf9c833a8361e2ae5645bc2de/torch/testing/_internal/common_nn.py#L5895-L5897

**Comment by jbschlosser** [comment]
> @jbschlosser I wonder if that is why the number of threads is set to one in `common_nn`:
> 
> https://github.com/pytorch/pytorch/blob/0561e104d97cb2abf9c833a8361e2ae5645bc2de/torch/testing/_internal/common_nn.py#L5895-L5897

Yeah good find, we should probably replicate that for the `ModuleInfo` tests.

**Comment by jbschlosser** [comment]
Finished in #68097
